### PR TITLE
Restructure and introduce read-write benchmark 

### PR DESCRIPTION
**Pull request by krishnangovindraj**
## What is the goal of this PR?
Implements the TypeDB benchmark using the simulation framework. This benchmark contains a 'minimal set of representative queries' to emulate a typical typedb workload.
It was developed to evaluate the effects of replacing RocksDB with SpeeDB on read & write performance, and to evaluate the effect of any future changes on performance.

## What are the changes implemented in this PR?
* Move the simulation framework into the `/framework` directory
* Introduce the read-write benchmark in the `read-write` directory 
* Apart from the minimal implementation of the simulation framework needed, the benchmark:
  * Implements `PersonAgent` as the only simulation agent, with:
    * An insert workload: `createPerson`
    * A match-insert workload: `createFriendship`
    * Four match workloads: `readAddressFromName`, `readFriendsOf`, `readFriendsOfFriends`, `readPersonsByPostCode`
    * See `/benchmark/README.md` for a longer description.
  * Introduces custom `ModelParameters` (configurable through the config file) of 
    * `personPerIteration`: determines how many `person` entities (and associated attributes) are inserted per run of `personAgent.createPerson`
    * `friendshipPerIteration`: determines how many `friendship` relations are inserted per run of `personAgent.createFriendship`
    * `postCodes`: determines how many unique `postCodes` exist, which determine the growth of the read in `readPersonsByPostCode`

### CODE REVIEW

File: framework/BUILD [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+load("@io_bazel_rules_kotlin//kotlin:jvm.bzl", "kt_jvm_library")
+load("@vaticle_dependencies//tool/checkstyle:rules.bzl", "checkstyle_test")
+
+kt_jvm_library(
+    name = "typedb_simulation",
```
Comment (flyingsilverfin) [inline framework/BUILD]: lets call this `typedb_simulation_framework` to match the names somewhat

File: read-write/BUILD [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+load("@io_bazel_rules_kotlin//kotlin:jvm.bzl", "kt_jvm_library")
+load("@vaticle_dependencies//tool/checkstyle:rules.bzl", "checkstyle_test")
+
+kt_jvm_library(
+    name = "storage-benchmark",
+    srcs = glob(["*.kt"]),
+#    resource_strip_prefix = "config/",
```
Comment (flyingsilverfin) [inline read-write/BUILD]: clean up if you can

File: read-write/BUILD [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+load("@io_bazel_rules_kotlin//kotlin:jvm.bzl", "kt_jvm_library")
+load("@vaticle_dependencies//tool/checkstyle:rules.bzl", "checkstyle_test")
+
+kt_jvm_library(
+    name = "storage-benchmark",
```
Comment (flyingsilverfin) [inline read-write/BUILD]: name consistently with package name

Comment (flyingsilverfin) [reply read-write/BUILD]: Also let's use `benchmark-read-write`

File: read-write/BUILD [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    srcs = glob(["*.kt"]),
+#    resource_strip_prefix = "config/",
+    deps = [
+        "//read-write/common",
+
+        # External Vaticle Dependencies
+        "//framework:typedb_simulation",
+        "//framework/typedb:typedb",
+        "//framework/common:common",
```
Comment (flyingsilverfin) [inline read-write/BUILD]: note: these are not `external` anymore, move up

File: read-write/BUILD [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        # External Maven Dependencies
+        "@maven//:com_google_code_findbugs_jsr305",
+        "@maven//:io_github_microutils_kotlin_logging_jvm",
+        "@maven//:org_slf4j_slf4j_api",
+    ],
+    data = [":schema.tql"]
+)
+
+java_binary(
+    name = "run",
```
Comment (flyingsilverfin) [inline read-write/BUILD]: poor name, please improve

File: read-write/README.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+bazel run //read-write:run -- --database=typedb --address=127.0.0.1:1729 --config=read-write/config/<config-file>.yml
+```
+
+Replace the address if you're not running the server on the same machine or using TypeDB's default port. 
+
+The <config-file> determines which queries will be run.
+* `ci-tests.yml`: Runs a few iterations to ensure everything is working properly (< 1 GB)
+* `read-write-benchmark.yml`: Performs (all) read & write queries described below (~10 GB)
+* `write-benchmark.yml`: Only performs write queries described below (~350 GB)
+* `large-data-benchmark.yml`: Performs all queries except `readAddressFromPostCode` for a lage number of iterations (~350 GB)
```
Comment (flyingsilverfin) [inline read-write/README.md]: typo

File: read-write/README.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
+# TypeDB read-write benchmark
+This benchmark contains a 'minimal set of representative queries' to emulate a typical typedb workload.
+It was developed to evaluate the effects of replacing RocksDB with SpeeDB on read & write performance.
+
+### Requirements 
+* Bazel v5.1.1
+* A running [typedb server](https://github.com/vaticle/typedb) to benchmark against.
```
Comment (flyingsilverfin) [inline read-write/README.md]: I'd make a note on the version of client-java it depends on and therefore which TypeDB server versions is it compatible with.

File: read-write/README.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+Replace the address if you're not running the server on the same machine or using TypeDB's default port. 
+
+The <config-file> determines which queries will be run.
+* `ci-tests.yml`: Runs a few iterations to ensure everything is working properly (< 1 GB)
+* `read-write-benchmark.yml`: Performs (all) read & write queries described below (~10 GB)
+* `write-benchmark.yml`: Only performs write queries described below (~350 GB)
+* `large-data-benchmark.yml`: Performs all queries except `readAddressFromPostCode` for a lage number of iterations (~350 GB)
+
+### Benchmark
```
Comment (flyingsilverfin) [inline read-write/README.md]: `## Benchmark Queries`

File: read-write/README.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  * then finds all friendships and the friend (other person) in each friendship
+* `readFriendsOfFriends`: pure-read
+  * selects a name at random (client-side)
+  * finds the person with the name
+  * then finds __all__ friendships and the friend in each friendship
+  * for each such 'other person', it finds __all__ their friendship relations and the associated friends. 
+* `readPersonsByPostCode`: pure-read
+  * selects a postCode at random (client-side)
+  * finds __all__ persons with the associated postCode.
+  * __Note__: Since the number of postCodes is fixed, this query scales linearly as more persons are inserted
```
Comment (flyingsilverfin) [inline read-write/README.md:53]: add a section `## Benchmark Architecture`

Let's add a section here which should outline how the benchmark operates in terms of multithreading, determinism, etc. (basically what are the partitions & how is parallelism achieved in relation to the above queries) This helps the new reader understand the some of the keywords such as 'partition' and whether things are deterministic or not, and how that's achieved.

File: read-write/StorageBenchmark.kt [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+import com.vaticle.typedb.client.api.TypeDBOptions
+import com.vaticle.typedb.client.api.TypeDBSession
+import com.vaticle.typedb.benchmark.readwrite.common.Context
+import com.vaticle.typedb.benchmark.framework.common.seed.RandomSource
+import com.vaticle.typedb.benchmark.framework.typedb.TypeDBClient
+import mu.KotlinLogging
+import java.nio.file.Paths
+
+class StorageBenchmark internal constructor(client: TypeDBClient, context: Context) :
```
Comment (flyingsilverfin) [inline read-write/StorageBenchmark.kt]: Rename

File: read-write/StorageBenchmark.kt [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    com.vaticle.typedb.benchmark.framework.typedb.TypeDBSimulation<Context>(client, context, AgentFactory(client, context)) {
+
+    private val LOGGER = KotlinLogging.logger {}
+    override val agentPackage = PersonAgent::class.java.packageName
+    override val name = "StorageBenchmark"
+
+    // TODO: Update this filepath
```
Comment (flyingsilverfin) [inline read-write/StorageBenchmark.kt]: hanging TODO - address or clean up if possible

File: read-write/common/BUILD [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+kt_jvm_library(
+    name = "common",
+    srcs = glob(["*.kt"]),
+    visibility = ["//visibility:public"],
+    deps = [
+        # External Vaticle Dependencies
+        "@vaticle_typedb_common//:common",
+        "//framework/common",
+        "//framework:typedb_simulation",
```
Comment (flyingsilverfin) [inline read-write/common/BUILD]: same thing here - please check all your BUILD files

File: read-write/config/BUILD [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+#
+
+load("@vaticle_dependencies//tool/checkstyle:rules.bzl", "checkstyle_test")
+
+package(default_visibility = ["//read-write:__pkg__"])
+
+filegroup(
+    name = "simulation_config",
+    srcs = [
+        "logback.xml",
```
Comment (flyingsilverfin) [inline read-write/config/BUILD:25]: logback is normally its own filegroup

File: read-write/config/ci-tests.yml [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+run:
+  randomSeed: 1
+  iterations: 5
+  partitions: 4
+  databaseName: "benchmark-on-ci"
+  recreateDatabase: true
+
+model:
+  personPerIteration: 10
```
Comment (flyingsilverfin) [inline read-write/config/ci-tests.yml]: change all these params to `perRun` in all the configs where relevant

**Review by flyingsilverfin — CHANGES_REQUESTED** [review]

File: read-write/README.md (hunk 9 above)
Comment (krishnangovindraj) [reply read-write/README.md:53]: ~~Shall we do this under framework and link?~~
done

File: dependencies/vaticle/repositories.bzl [hunk 15]
```diff
@@ -21,14 +21,14 @@ def vaticle_dependencies():
...
-        commit = "20566a03cf11b821e9f1b88698f3099b39616ae6",  # sync-marker: do not remove this comment, this is used for sync-dependencies by @vaticle_dependencies
+        commit = "6a8c31f57b8f7e10d63f00ef709f80c301aaedaf",  # sync-marker: do not remove this comment, this is used for sync-dependencies by @vaticle_dependencies
     )
 
 def vaticle_typedb_client_java():
     git_repository(
         name = "vaticle_typedb_client_java",
         remote = "https://github.com/vaticle/typedb-client-java",
-        commit = "c2d0fd5b0d5e5629f8166b9a3bb5d9379d648da1",  # sync-marker: do not remove this comment, this is used for sync-dependencies by @vaticle_typedb_client_java
+        tag = "2.16.1",  # sync-marker: do not remove this comment, this is used for sync-dependencies by @vaticle_typedb_client_java
```
Comment (flyingsilverfin) [inline dependencies/vaticle/repositories.bzl]: can we use `2.18.0` now?

File: framework/common/BUILD [hunk 16]
```diff
@@ -32,7 +32,7 @@ kt_jvm_library(
         "@maven//:io_github_microutils_kotlin_logging_jvm",
         "@maven//:org_apache_commons_commons_csv",
         "@maven//:org_slf4j_slf4j_api",
-        "@maven//:org_yaml_snakeyaml",
+#        "@maven//:org_yaml_snakeyaml",
```
Comment (flyingsilverfin) [inline framework/common/BUILD]: ?

File: read-write/README.md [hunk 17]
```diff
@@ -0,0 +1,113 @@
+## TypeDB read-write benchmark
+This benchmark contains a 'minimal set of representative queries' to emulate a typical typedb workload.
+It was developed to evaluate the effects of replacing RocksDB with SpeeDB on read & write performance,
+and to evaluate the effect of any future changes on performance.
```
Comment (flyingsilverfin) [inline read-write/README.md:5]: newline

**Review by flyingsilverfin — APPROVED** [review]
Approved with 3 minor comments, please address then merge!

**Pull request merged**